github-webhook-util/Main.hs

```diff
@@ -0,0 +1,274 @@
+{-# OPTIONS -Weverything -Wno-implicit-prelude #-}
+
+module Main (main) where
+
+import           Control.Monad.Operational    (Program,
+                                               ProgramViewT ((:>>=), Return),
+                                               singleton, view)
+import           Control.Monad.Trans.Except   (ExceptT (ExceptT), runExceptT)
+import qualified Data.HashMap.Strict          as HM
+import           Data.Int                     (Int64)
+import qualified Data.Map                     as Map
+import           Data.Maybe                   (fromJust, fromMaybe)
+import           Data.Monoid                  ((<>))
+import           Data.Reflection              (Given, give, given)
+import           Data.String                  (fromString)
+import           Data.String.Encode           (encodeString)
+import           Data.Time                    (UTCTime, diffUTCTime)
+import           Data.Vector                  (Vector)
+import qualified Data.Vector                  as V
+import           Database.InfluxDB            (Database, Field (FieldInt),
+                                               Line (Line), LineField, write,
+                                               writeParams)
+import           Database.InfluxDB.Types      (Key (Key), Measurement)
+import qualified GitHub                       as GH
+import qualified GitHub.Data.Id               as GH
+import           GitHub.Data.Webhooks.Events  (StatusEvent, StatusEventState (StatusFailureState, StatusSuccessState),
+                                               evStatusCommitSha, evStatusRepo,
+                                               evStatusState)
+import           GitHub.Data.Webhooks.Payload (HookSimpleUser (HookSimpleUser),
+                                               whRepoName, whRepoOwner,
+                                               whUserLogin)
+import           Network.HTTP.Client          (Manager, newManager)
+import           Network.HTTP.Client.TLS      (tlsManagerSettings)
+import           System.Environment           (getProgName, lookupEnv)
+
+import           Control.Monad.IO.Class       (liftIO)
+import qualified Data.Aeson                   as AE
+import qualified Data.ByteString.Char8        as C8
+import           Network.Wai.Handler.Warp     (run)
+import           Servant                      ((:>), Context ((:.)))
+import qualified Servant                      as S
+import qualified Servant.GitHub.Webhook       as S
+
+type GithubMonad a (k :: GH.RW) = Program (GH.Request k) a
+type HasGHAuth = Given (Manager, GH.Auth)
+
+data AccumulatorValue = AccumulatorValue {
+    avCommit      :: GH.Commit
+  , avStatuses    :: Vector GH.Status
+  , avFirstStatus :: Maybe UTCTime
+  } deriving Show
+
+data Accumulator = Accumulator {
+    ghStatusMap :: HM.HashMap (GH.Name GH.Commit) AccumulatorValue
+  } deriving Show
+
+getAuth :: IO (Maybe GH.Auth)
+getAuth = do
+    token <- lookupEnv "GITHUB_TOKEN"
```

Comment:
Let's document the least privileges required for this OAuth token and ensure the one we use is configured that way.